synapse/storage/data_stores/main/end_to_end_keys.py

```diff
@@ -330,6 +332,105 @@ def get_e2e_cross_signing_key(self, user_id, key_type, from_user_id=None):
             from_user_id,
         )
 
+    def _get_e2e_cross_signing_keys_bulk_txn(
+        self, txn, users, key_type, from_user_id=None
+    ):
+        """Returns the cross-signing keys for a set of users.
+
+        Args:
+            txn (twisted.enterprise.adbapi.Connection): db connection
+            users (iterable[str]): the users whose keys are being requested
+            key_type (str): the type of keys that are being requested: either 'master'
+                for a master key, 'self_signing' for a self-signing key, or
+                'user_signing' for a user-signing key
+            from_user_id (str): if specified, signatures made by this user on
+                the keys will be included in the result
+
+        Returns:
+            dict mapping from user ID to key data.  If a user's cross-signing key
+                was not found, their user ID will not be in the dict.
+        """
+        sql = (
+            "SELECT k.user_id, k.keydata, k.stream_id "
+            "  FROM e2e_cross_signing_keys k"
+            "  JOIN (SELECT user_id, MAX(stream_id) AS stream_id"
+            "          FROM e2e_cross_signing_keys"
+            "         WHERE keytype = ?"
+            "         GROUP BY user_id) s"
+            " USING (user_id, stream_id)"
+            " WHERE k.user_id IN (%s) AND k.keytype = ?"
```

Comment:
Can you put this in a multiline string literal please (i.e.` """ ... """`)? That its a lot easier to see what's going on and to c+p